Userモデルを作成

tatsukishiga opened the pull request:
モデルを作成し、パスワード暗号化時に必要な文とバリデーション、アソシエーションを追加した。

ArakawaShintaro commented on app/controllers/users_controller.rb:
```diff
@@ -0,0 +1,25 @@
+class UsersController < ApplicationController
+  def new
```
ログイン、新規登録、ログアウトに関係するアクションのみを定義しましょう

ArakawaShintaro replied:
test

tatsukishiga committed:
Userモデルを作成

モデルを作成し、パスワード暗号化時に必要な文とバリデーション、アソシエーションを追加した。

tatsukishiga committed:
ユーザーのsession機能を作成

ユーザーコントローラーを作成し、新規登録、ログイン、ログアウト機能を追加。アプリケーションコントローラーにはsessionに関係するメソッドを3つ追加。ビューではログインフォームと新規登録画面を追加し、各リンクを作成した。実行後の通知、エラーメッセージも表示されるようにした。

tatsukishiga committed:
ユーザー詳細ページを追加

詳細ページ（マイページ）を追加しリンクも書いた。

tatsukishiga committed:
ユーザーを編集ページを追加

ユーザー詳細ページから編集ページに移れるようにした。

tatsukishiga committed:
ユーザーの更新機能を追加

ユーザー編集ページで変更した情報をアップデート出来るようにした。またプロフィールのdefault画像も設定した。

tatsukishiga committed:
ユーザーの削除機能を追加

ユーザー編集ページからアカウントを削除出来る機能を追加した。

ArakawaShintaro commented on app/controllers/application_controller.rb:
```diff
@@ -1,2 +1,21 @@
 class ApplicationController < ActionController::Base
+  before_action :set_current_user
+# ログイン中のユーザーを取得する
```
無駄なコメントアウトは消してもいいと思います
これくらいの処理であれば何をしているのかコードを読めばわかるので

ArakawaShintaro replied:
他のコメントアウトも全てそうです

tatsukishiga replied:
了解です。コメントアウトを削除します！

ArakawaShintaro commented on app/controllers/application_controller.rb:
```diff
@@ -1,2 +1,21 @@
+  def forbid_login_user
+    if @current_user
+      flash[:notice] = "すでにログインしています"
+      redirect_to("/posts/index")
```
なぜposts/indexに飛ばしているのでしょう？
posts/indexってありましたっけ？
ないのであれば、ここはひとまずリダイレクト先を `#` などで代用しておけばOKだと思います。

tatsukishiga replied:
postsではなく、topicsの間違いでした；；
すでにログインしている場合はトピック一覧ページに飛ばしたいので
指摘いただいた点を修正します。

ArakawaShintaro commented on test/controllers/users_controller_test.rb:
```diff
@@ -0,0 +1,29 @@
+require 'test_helper'
```
テストは今回はいらないので消しましょう

tatsukishiga replied:
はい、消します！

ArakawaShintaro commented on test/fixtures/users.yml:
```diff
@@ -0,0 +1,15 @@
+# Read about fixtures at http://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
```
これもいらないです

tatsukishiga replied:
消します！

ArakawaShintaro commented on test/models/user_test.rb:
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
```
これも

tatsukishiga replied:
消します！

ArakawaShintaro reviewed (commented):
今回は直さなくていいですが、showとかeditとかこのpull-reqでいらないものが多々見受けられますね

tatsukishiga committed:
session周りを修正

不要なコメントアウトを削除。すでにログインしている時に飛ばすページを修正。またテスト関係も今回不要なので削除した。

tatsukishiga committed:
ユーザー詳細に投稿を表示

ユーザー詳細ページにトピックへの投稿内容を表示されるようにした。

tatsukishiga committed:
ユーザー編集ページ修正

プロフィールの<p>タグを追加した。

ArakawaShintaro commented on app/controllers/users_controller.rb:
```diff
@@ -0,0 +1,80 @@
+      File.binwrite("public/profile_images/#{@user.profile_image}", image.read)
+    end
+
+    if @user.save
```
updateメソッドがあるので、そちらを使用してみましょう

tatsukishiga replied:
かしこまりました

ArakawaShintaro commented on app/views/users/show.html.erb:
```diff
@@ -0,0 +1,24 @@
+      <img src="<%= "/profile_images/#{@user.profile_image}" %>">
+      <h2><%= @user.name %></h2>
+      <textarea name="profile"><%= @user.profile %></textarea>
+      <% if @user.id == @current_user.id %>
```
helperを使ってメソッド化してみましょう

tatsukishiga replied:
了解です

ArakawaShintaro commented on config/routes.rb:
```diff
@@ -1,3 +1,13 @@
-  # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
+  # users
+  get 'signup' => 'users#new'
+  post 'users/create' => 'users#create'
```
users周りはresourcesメソッドを使って定義しましょう

tatsukishiga replied:
了解です

ArakawaShintaro commented on db/migrate/20190302031555_create_users.rb:
```diff
@@ -0,0 +1,13 @@
+  def change
+    create_table :users do |t|
+      t.string :name
+      t.string :email
```
制約をつけましょう

rollback -> 変更 -> migrate

tatsukishiga replied:
MBPro-2:board3 shigatatsuki$ rails db:migrate:status
database: board3_development
 Status   Migration ID    Migration Name
--------------------------------------------------
   up     20190302031555  Create users
   up     20190302231928  ********** NO FILE **********
   up     20190315125041  Change column user

真ん中の「NO FILE」はトピックブランチの方で使用しているものなのですが、こちらが邪魔をしてロールバックできなかったため、新たにmigrationファイルを生成して制約をつけました。

tatsukishiga committed:
指摘箇所修正

1.URLで直打された場合の対策。
2.飛ばすURLにpathを使用。
3.form_tagをform_withに変更。
4.Resourcesを使用してルーティングを定義
5.ビュー内での処理をhelperを使用してメソッド化。
6.updateアクションを追加。
7.新たにmigrationファイルを生成しdb側にバリデーションを追加。
8.メールアドレスのフォーマットチェックをバリデーションに追加。

ArakawaShintaro commented on app/controllers/users_controller.rb:
```diff
@@ -0,0 +1,88 @@
+    if @user.save
+      session[:user_id] = @user.id
+      flash[:notice] = "ユーザー登録が完了しました"
+      redirect_to("/users/#{@user.id}")
```
名前付きパスでできるところはすべてそれでやりましょう

tatsukishiga committed:
名前付きpathに変更

名前付きpathに出来るところを全て直した。renderについてはこの記述が一番わかりやすくシンプルかなと思いましたのでこちらで定義させていただきました。

ArakawaShintaro reviewed (approved):
LGTM

tatsukishiga committed:
Merge branch 'master' into user_function
